### PR TITLE
Xbutil2 u30 enhancement 

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
CR : https://jira.xilinx.com/browse/CR-1087398

XRT needs to monitor more metrics for the edge stack running on U30 PS. We should at the minimum enhance xbutil query to 
report the following:

Xbutil2 Enhancement
1. Memory Report 
Show all buffers information in memory report
- xclGetHostBO
- xclMapBO
- xclUnmapBO
- xclFreeBO

2. SKD report 
Show all the skd processes running
- Show how many times a SKD triggered 
- Metrics on how many times SKDs have crashed or have returned nonzero error code

### CODE REVIEW

File: src/runtime_src/core/common/drv/include/kds_core.h [hunk 1]
```diff
@@ -98,12 +98,27 @@ struct kds_client {
...
+};
+
 struct kds_scu_mgmt {
 	struct mutex		  lock;
 	int			  num_cus;
 	u32			  status[MAX_CUS];
 	u32			  usage[MAX_CUS];
 	char			  name[MAX_CUS][32];
+	struct scu_usages	  sc_usages[MAX_CUS];
+	struct sk_mem_stats	  sk_mem_stats;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/drv/include/kds_core.h:121]: As this structure name is already contains scu, better to use member names like usage_stats. and mem_stats

File: src/runtime_src/core/common/drv/kds_core.c [hunk 2]
```diff
@@ -93,13 +93,39 @@ ssize_t show_kds_scustat_raw(struct kds_sched *kds, char *buf)
...
+{
+	struct kds_scu_mgmt *scu_mgmt = &kds->scu_mgmt;
+	struct sk_mem_stats *mem_stat;
+
+	mutex_lock(&scu_mgmt->lock);
+
+	mem_stat = &scu_mgmt->sk_mem_stats;
+	
+	/* Check if this is present. Non Scu case fail here */
+	if (!scu_mgmt && !mem_stat)
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/drv/kds_core.c:116]: why is this check here? we are anyway taking address of a member variable. ?

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 3]
```diff
@@ -324,11 +324,15 @@ static inline void read_ert_stat(struct kds_command *xcmd)
 	mutex_lock(&kds->scu_mgmt.lock);
 	/* Skip header and FPGA CU stats. off_idx points to PS kernel stats */
 	off_idx = 4 + num_cu;
-	for (i = 0; i < num_scu; i++)
-		kds->scu_mgmt.usage[i] = ecmd->data[off_idx + i];
+	for (i = 0; i < num_scu; i++) {
+		kds->scu_mgmt.usage[i] = ecmd->data[off_idx++];
+		kds->scu_mgmt.sc_usages[i].succ_cnt = ecmd->data[off_idx++];
+                kds->scu_mgmt.sc_usages[i].err_cnt = ecmd->data[off_idx++];
+                kds->scu_mgmt.sc_usages[i].crsh_cnt = ecmd->data[off_idx++];
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:331]: indentation

File: src/runtime_src/core/tools/common/ReportCu.cpp [hunk 4]
```diff
@@ -161,7 +161,9 @@ populate_cus_new(const xrt_core::device *device)
 
   try {
     cu_stats  = xrt_core::device_query<qr::kds_cu_stat>(device);
+    std::cout << cu_stats.size() << std::endl;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportCu.cpp:164]: remove cout

File: src/runtime_src/core/tools/common/ReportCu.cpp [hunk 5]
```diff
@@ -161,7 +161,9 @@ populate_cus_new(const xrt_core::device *device)
 
   try {
     cu_stats  = xrt_core::device_query<qr::kds_cu_stat>(device);
+    std::cout << cu_stats.size() << std::endl;
     scu_stats = xrt_core::device_query<qr::kds_scu_stat>(device);
+    std::cout << scu_stats.size() << std::endl;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportCu.cpp:166]: remove cout

File: src/runtime_src/core/tools/common/ReportCu.cpp [hunk 6]
```diff
@@ -203,6 +205,9 @@ populate_cus_new(const xrt_core::device *device)
     ptCu.put( "base_address",   "0x0");
     ptCu.put( "usage",          stat.usages);
     ptCu.put( "type", enum_to_str(cu_type::PS));
+    ptCu.put( "succ_return",    stat.succ_cnt);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportCu.cpp:208]: better to use full names like: successful_return etc

File: src/runtime_src/core/tools/common/ReportMemory.cpp [hunk 7]
```diff
@@ -225,6 +244,20 @@ populate_memtopology(const xrt_core::device * device, const std::string& desc)
...
+  std::stringstream ss(mm_buf[map->m_count + 1]);
+  if (!ss.str().empty()) {
+    uint64_t hboCnt, mapboCnt, unmapboCnt, freeboCnt;
+  
+    ss >> hboCnt >> mapboCnt >> unmapboCnt >> freeboCnt;
+    ptSkMem.put("sk_hbo", hboCnt);
+    ptSkMem.put("sk_mapbo", mapboCnt);
+    ptSkMem.put("sk_unmapbo", unmapboCnt);
+    ptSkMem.put("sk_freebo", freeboCnt);
+    pt.add_child(std::string("board.memory.sk_memories"), ptSkMem);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportMemory.cpp:259]: sk_memory_usage would be better instead of sk_memories

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]

**Comment by chvamshi-xilinx** [comment]
@saifuddin-xilinx , Please make sure your changes works with old device XRT and new host XRT and vice-versa

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 8]
```diff
@@ -324,11 +324,15 @@ static inline void read_ert_stat(struct kds_command *xcmd)
...
+        }
 
 	/* off_idx points to PS kernel status */
-	off_idx += num_scu + num_cu;
+	off_idx += num_cu;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:335]: This calculation doesnt seem to be correct. Please check

**Comment by chvamshi-xilinx** [comment]
Can you also add a sample output of both CU and memory stats

**Comment by xilinxgitops** [comment]
retest this please

File: src/runtime_src/core/common/drv/include/kds_core.h (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_core.h]: Updated the name as suggested.

File: src/runtime_src/core/common/drv/kds_core.c (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c]: Removed this check.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c]: Solved

File: src/runtime_src/core/tools/common/ReportCu.cpp (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportCu.cpp]: Done

File: src/runtime_src/core/tools/common/ReportCu.cpp (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportCu.cpp]: Done

File: src/runtime_src/core/tools/common/ReportMemory.cpp (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportMemory.cpp]: Updated

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:335]: Actually there the communication between Xocl and Zocl through packet payload. 
The payload structure ads follows for this structure.
+------------------------+
|     custat version          |  4 (32bit)
|     git version               |
|     # CQ slots               |
|     # CUs                      |
+------------------------+
| individual CU stat        |  num_cus
|        Total usages          |
+------------------------+
| individual SCU stat       | num_scu
|        Total usages          |
|		 Succ Count    |    
|		 Error Count    |
|		 Crash Count  |
+------------------------+
| individual CU status     | num_cus
|                                      |
+------------------------+
| individual SCU status   | num_scu
|                                      |
+------------------------+
| SK Memory status       | 
|     Hbo count               |
|     Map count               |
|     Unmap count           |
|     Free count               |
+------------------------+
|  slot status                   |  num_slots
+------------------------+

Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:335]: As per the offline discussion with you this calculation is fine.

**Review by mamin506 — APPROVED** [review]
Overall, this looks good to me.
Since you have changed the xbutil tools output. Could you provide sample output for text/json? You can ref #5277

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 9]
```diff
@@ -868,6 +879,9 @@ void zocl_free_host_bo(struct drm_gem_object *gem_obj)
 	drm_gem_object_release(gem_obj);
 
 	kfree(&zocl_bo->cma_base);
+			
+	/* Update softkernel memory stats */
+        zocl_update_sk_mem_stat(zdev, 1, ZOCL_MEM_STAT_TYPE_FREEBO);
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c:884]: indent

File: src/runtime_src/core/edge/skd/sk_daemon.cpp [hunk 10]
```diff
@@ -204,11 +205,13 @@ static void softKernelLoop(char *name, char *path, uint32_t cu_idx)
     }
 
     /* Reg file indicates the kernel should not be running. */
-    if (!(args_from_host[0] & 0x1))
+    if (!(args_from_host[0] & 0x1))
```
Comment (larry9523) [inline src/runtime_src/core/edge/skd/sk_daemon.cpp:208]: extra space at the end of line

**Review by larry9523 — CHANGES_REQUESTED** [review]
What if a new host XRT is running against old shell without the enhancement? I believe there is unused space in the this ERT_CU_STAT header that can be used as a version control

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good.

File: src/runtime_src/core/include/ert.h [hunk 11]
```diff
@@ -596,6 +596,13 @@ uint32_t ert_base_addr = 0;
 # define ERT_INTC_CU_96_127_ADDR           0x3000
 #endif
 
+	
+/**
+ * This defines ERT packet version which need to sync between Host and ERT 
+ */
+#define ERT_PACKET_VERSION_0		  0x51a10000
+#define ERT_PACKET_VERSION_1		  0x51a10001
```
Comment (larry9523) [inline src/runtime_src/core/include/ert.h:604]: Thanks for adding the version check! I suggest to rename it to ERT_CUSTAT_VERSION other than the packet version.

**Review by larry9523 — CHANGES_REQUESTED** [review]
Please consider changing the Macro name.

Other than that, make sure xbutil runs OK on old and new device xrt. What is the behavior if old host xrt runs against new device xrt? We don't need to support it but we may want to understand it and have proper document. Make sure there is no crash or other bad thing happen.

Thanks!

File: src/runtime_src/core/tools/common/ReportMemory.cpp [hunk 12]
```diff
@@ -389,6 +422,33 @@ ReportMemory::writeReport( const xrt_core::device* /*_pDevice*/,
...
+      for (auto& v : _pt.get_child("mem_topology.board.memory.sk_memory_usage",empty_ptree)) {
+        if (v.first == "sk_hbo") {
+          hbo_count = v.second.get_value<std::string>();
+        } else if (v.first == "sk_mapbo") {
+          mapbo_count = v.second.get_value<std::string>();
+        } else if (v.first == "sk_unmapbo") {
+          unmapbo_count = v.second.get_value<std::string>();
+        } else if (v.first == "sk_freebo") {
+          freebo_count = v.second.get_value<std::string>();
+	}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportMemory.cpp:442]: indentation

**Review by AShivangi — COMMENTED** [review]
This looks good. Can you also add the output for the memory report to the description?

With these changes, is there any change in the json? Can you highlight what those are?
Thanks!

File: src/runtime_src/core/tools/common/ReportCu.cpp (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportCu.cpp]: Done

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Fixed.

File: src/runtime_src/core/edge/skd/sk_daemon.cpp (hunk 10 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/skd/sk_daemon.cpp]: Fixed

File: src/runtime_src/core/include/ert.h (hunk 11 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/include/ert.h]: Updated. Thanks.

**Review by larry9523 — APPROVED** [review]

**Review by houlz0507 — APPROVED** [review]

**Comment by saifuddin-xilinx** [comment]
This PR has been created on origin/u30 branch. But currently U30 branch is upgraded to 2021.1 branch. 
Hence created a new PR targeted to 2021.1 and closing this PR. 
New PR: 
https://github.com/Xilinx/XRT/pull/5515

**Pull request closed without merging**